include/ginkgo/core/base/collection.hpp

```diff
@@ -0,0 +1,200 @@
+        *this = other;
+    }
+
+    array(std::shared_ptr<const Executor> exec, array&& other) : array(exec)
+    {
+        *this = std::move(other);
+    }
+
+    array(const array& other) { *this = other; }
+
+    array& operator=(const array& other)
+    {
+        if (this != &other) {
+            buffer_ = other.buffer_;
+            offsets_ = other.offsets_;
+            elems_.resize(other.elems_.size());
+            for (size_type i = 0; i < elems_.size(); ++i) {
+                elems_[i] = make_array_view(
+                    buffer_.get_executor(), other[i].get_size(),
+                    buffer_.get_data() +
+                        std::distance(other.buffer_.get_const_data(),
+                                      other[i].get_const_data()));
+            }
+        }
+        return *this;
+    }
+
+    // can't be noexcept, because array is not noexcept
+    array(array&&) = default;
+
+    array& operator=(array&&) = default;
+
+    template <typename SizeType = std::size_t>
+    array(std::shared_ptr<const Executor> exec,
+          const std::vector<SizeType>& sizes)
+        : array(gko::array<T>(exec,
+                              std::accumulate(sizes.begin(), sizes.end(), 0)),
+                sizes)
+    {}
+
+    template <typename SizeType = std::size_t>
+    array(gko::array<T> buffer, const std::vector<SizeType>& sizes)
+        : buffer_(std::move(buffer))
+    {
+        auto exec = buffer_.get_executor();
+        offsets_ = gko::array<size_type>(exec->get_master(), sizes.size() + 1);
+        offsets_.fill(0);
+        std::partial_sum(sizes.begin(), sizes.end(), offsets_.get_data() + 1);
+        buffer_.resize_and_reset(offsets_.get_data()[sizes.size()]);
+        // can't use emplace/push_back because then the view will get copied
+        // into owning arrays
+        elems_.resize(sizes.size());
+        for (size_type i = 0; i < sizes.size(); ++i) {
+            elems_[i] = make_array_view(
+                exec, sizes[i], buffer_.get_data() + offsets_.get_data()[i]);
+        }
+        offsets_.set_executor(exec);
+    }
+
+    reference operator[](size_type i) { return elems_[i]; }
+
+    const_reference operator[](size_type i) const { return elems_[i]; }
+
+    [[nodiscard]] size_type size() const { return elems_.size(); }
+
+    iterator begin() { return elems_.data(); }
+
+    const_iterator begin() const { return elems_.data(); }
+
+    iterator end() { return begin() + elems_.size(); }
+
+    const_iterator end() const { return begin() + elems_.size(); }
+
+    [[nodiscard]] bool empty() const { return elems_.empty(); }
+
+    reference get_flat() { return buffer_; }
+
+    const_reference get_flat() const { return buffer_; }
+
+    const gko::array<size_type>& get_offsets() const { return offsets_; }
+
+    std::shared_ptr<const Executor> get_executor() const
+    {
+        return buffer_.get_executor();
+    }
+
+private:
+    gko::array<T> buffer_;
+    gko::array<size_type> offsets_;
```

Comment:
Considering #430, I would prefer if we used int64 or even a generic integer type.